### ISSUE TITLE
IBM 5155 console keyboard 

### ISSUE THREAD

**Issue by Mellvik** [Mellvik/TLVC#66]
Not seen on other hardware, but on the IBM 5155, the "XT Portable", the keyboard doesn't work with TLVC. 

Works fine with DOS, serial io is not affected. Also works fine with COMPAQ Portable Plus, also an XT class machine. Since the 5155  has a standard XT mainboard, this problem presumably hits all IBM PC and XT models.

**Comment by vkoskiv** [comment]
Confirmed on my 5150 (+5160 expansion unit), 83-key keyboard doesn't seem to work on the login prompt.

**Comment by Mellvik** [comment]
It turns out this problem appears only when the root volume is FD. Booting from HD, or even from FD with root=<some HD volume> in bootopts, and we're good. So it seems the sequence of initialization is key. 
When starting from HD, using the FD doesn't affect the keyboard, it still works. In short it seems like if the FD is accessed (driver opened) before the console (keyboard really) is completely initialized, the kbd locks up. Maybe an interrupt mask.

@ghaerr, would you possibly have an idea about where to look for this one?

**Comment by ghaerr** [comment]
> In short it seems like if the FD is accessed (driver opened) before the console (keyboard really) is completely initialized, the kbd locks up.

If setting the root volume in /bootopts causes the problem to go away, then I would disagree with your conclusion above, as changing the root volume in /bootopts doesn't reorder any console initialization.

What does differ between actual BIOS boot to a drive number versus /bootopts etc options is that on BIOS boot, register DL contains the BIOS drive number to boot from. The boot loader then uses this information to boot from the specified device. But I don't think that sounds like the issue.

Other possible issues are that the A20 line is turned on/off/tested at boot when XMS is enabled - and A20 fiddling uses the keyboard controller. You can look at this code in elks/arch/i86/lib/unreal.S, functions enable_a20_gate, etc. That file has a couple of options to enable/disable A20, and by default uses an INT 15h function, which possibly isn't implemented on your system, or something strange? The top of the file documents the various options, if this is even the issue.

There may be something funky with the actual communications to the keyboard controller chip, but I don't remember offhand which bits are masked and rerouted, etc. It might be worth looking at the 8042  in the IBM XT documentation, so we can have a more technical discussion.

Of course, turn scan code keyboard off (which should be automatic with BIOS console, but not sure). With the scan code keyboard disabled, IIRC the kernel polls the keyboard using a BIOS INT16h interrupt to check whether there's a character received - this is entirely different than the IRQ 1 interrupt occurring with the scan code driver.

**Comment by Mellvik** [comment]
> > In short it seems like if the FD is accessed (driver opened) before the console (keyboard really) is completely initialized, the kbd locks up.
> 
> If setting the root volume in /bootopts causes the problem to go away, then I would disagree with your conclusion above, as changing the root volume in /bootopts doesn't reorder any console initialization.

No, it does not. It's using HD (as opposed to floppy) that cause the problem to go away. The /bootopts root= example was there to demonstrate that the boot device has nothing to do with it, only the root device. 

> 
> There may be something funky with the actual communications to the keyboard controller chip, but I don't remember offhand which bits are masked and rerouted, etc. It might be worth looking at the 8042 in the IBM XT documentation, so we can have a more technical discussion.

Yes, it looks like that's the way we're heading. It's interesting though that the root device determines whether it works or not, so it does look like the sequence of things (initialization) is part of it.

> 
> Of course, turn scan code keyboard off (which should be automatic with BIOS console, but not sure). With the scan code keyboard disabled, IIRC the kernel polls the keyboard using a BIOS INT16h interrupt to check whether there's a character received - this is entirely different than the IRQ 1 interrupt occurring with the scan code driver.

I was looking onto this last night, only to find that kbd driver selection, with or without the scancode 'driver', don't change anything. Seems I need a deeper dive into this. My other XT - the compaq portable plus - called in sick (PS problem), but I seem to remember it did not show the problem. Neither does the V20 system.

**Comment by Mellvik** [comment]
Completely inadvertently I discovered that the keyboard works when BIOS IO (disk & floppy) is active (or the other way around, when direct floppy & hd are active, no keyboard). That narrows down the options (culprits) quite a bit.

**Comment by Mellvik** [comment]
In what ended up in another rabbit hunt spanning several days, this one was finally nailed. Well, not completely, but a workaround - and FWIW, the previous statement is wrong, BIOS or drivers have nothing to do with it.

What happens is that when the `clock` utility probes for CMOS availability in order to check whether the regular (AT type) RTC is available, the keyboard locks up. A single `outb()` to the CMOS control port (0x70) is all it takes. A reboot is required to release the 'lock' on the keyboard.

I haven't looked into why this is happening and the architecture (i.e. XT) cannot be used to prevent `clock` from accessing this port because many XT class systems have 3rd party CMOS clocks these days. So the workaround is to use the existing `-A` option in `clock` which will now (PR coming) prevent `cmos_probe()` from being run.

Not ideal, but a fairly simple solution to a really annoying problem. @vkoskiv - please test it on your system whenever you have a chance.

**Comment by ghaerr** [comment]
I think I've figured it out - the PC/XT doesn't decode "odd" upper nibble ports, meaning that writing to port 0x70/71 is the same as writing to 0x60/61. And the keyboard is at 0x60/61. So your probe code might be directly disabling the keyboard when writing port 0x71. 

[Source: IBM Tech Reference pg 2-23](https://www.minuszerodegrees.net/manuals/IBM/IBM_5150_Technical_Reference_6025005_AUG81.pdf).

Here's the XT I/O map (pg 2-23):
![Image](https://github.com/user-attachments/assets/2a6595b2-8a96-46c6-baa2-be0771a18220)

Here's the XT keyboard I/O map at 0x60-61: (pg 2-24):
![Image](https://github.com/user-attachments/assets/3c2fcec0-1efe-4726-92df-ee00b0935c8b)

Perhaps there's a way to rewrite your cmos_probe() such that the keyboard bit isn't changed, by reading the ports value first before writing or something like that?

**Comment by Mellvik** [comment]
Great research @ghaerr, this explains the problem completely. And enables experimentation towards a real fix instead of a workaround.

What's happening is that a single read from the CMOS (and because of this 'anomaly' the keyboard cmd port) is all it takes to lock it up. The read seems to set a state in the keyboard, IOW it's part of a sequence that must be completed in order for the keyboard to work. If the probe can complete that sequence, the keyboard should be OK. I will look into that.

Thanks.

**Comment by Mellvik** [comment]
Fixed in #137 .
